New serverless pattern - EventBridge Scheduler to invoke Step Functions to get the List of Inactive Lambda functions (#2533)
aws-samples/serverless-patterns: eventbridge-stepfunction-lambda-getfunctionstatus/README.md

> # EventBridge scheduler to invoke step function to get the list of inactive lambda functions

Reviewer: ```suggestion
# Amazon EventBridge Scheduler to invoke an AWS Step Function to get the list of inactive AWS Lambda functions
```

> This pattern will create a one time schedule in EventBridge scheduler with state machine as target & invoke the lambda ListFunctions & GetFunction API using the SDK integration in step functions and publish the list of inactive functions to SNS Topic.

Reviewer: ```suggestion
This pattern will create an Amazon EventBridge Scheduler rule targeting an AWS Step Function state machine. The Step Function invokes an AWS Lambda using `ListFunctions` and `GetFunctions` API using the SDK integration in Step Functions. It publishes the list of inactive functions to an Amazon SNS Topic.
```

## Deployment Instructions

1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
    ``` 
    git clone https://github.com/aws-samples/serverless-patterns
    ```
2. Change directory to the pattern directory:
    ```
    cd eventbridge-stepfunction-lambda-getfunctionstatus
    ```
3. Build the dependencies:
    ```
    sam build
    ```
4. From the command line, use AWS SAM to deploy the AWS resources for the pattern as specified in the template.yml file:
    ```
    sam deploy --guided
    ```
5. During the prompts:

    * Enter **stack name**.
    * Enter desired **AWS Region**.
    * Enter your **Email Address** (e.g. abcd@xyz.com) for the EmailAddress parameter.
    * Allow SAM CLI to create IAM roles with the required permissions.

    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.

6. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.

> 7. **Imp** : Once the SNS topic & subscription is created, verify the subscription by clicking the 'Confirm subscription' link received on your email. 

Reviewer: ```suggestion
7. **Important** : Once the SNS topic & subscription is created, verify the subscription by clicking the 'Confirm subscription' link received on your email. 
```
"Imp" means "important"?

## How it works

> 1. When the stack is deployed it creates a state machine, an eventBridge scheduler,a SNS topic with email subscription, a Lambda Function and required IAM roles. 

Reviewer: ```suggestion
1. When the stack is deployed it creates a state machine, an Event Bridge Scheduler, an SNS topic with email subscription, a Lambda function and required IAM roles. 
```

> 2. Event Bridge scheduler invokes the State machine on defined schedule, Default schedule invokes the state machine daily at 00::00 UTC with the below mentioned payload. Modify the schedule as per your requirement.

Reviewer: ```suggestion
2. Event Bridge Scheduler invokes the state machine with a defined schedule. The default schedule invokes the state machine daily at 00:00 UTC with the below mentioned payload. Modify the schedule as per your requirement.
```

    ```json
    {
        "Function_ARN_List_existing": {
        "lambda_arn_list_combined": []
        },
        "NextMarker": null
    }
    ```

> 3) In state machine, the 1st step is to get the list of all lambda functions in the current region using the ListFunctions API, ListFunctions API is called multiple times as it returns max 50 functions list in a single API call along with NextMarker pagination parameter. Using NextMarker token state machine retrieves list of all the function ARN's.

Reviewer: ```suggestion
3) In the state machine, the first step is to get the list of all Lambda functions in the current region using the ListFunctions API. The ListFunctions API is called multiple times as it returns a maximum of 50 function listed in a single API call, along with a NextMarker pagination parameter. Using the NextMarker token, the state machine retrieves a list of all the function ARNs.
```

> 4) Lambda Invoke Combine data Step : This a custom lambda function python script to create a combined json array of Lambda function ARN's retrieved in step 1.

Reviewer: ```suggestion
4) Lambda Invoke Combine data Step : This a custom Lambda function python script to create a combined JSON array of Lambda function ARNs retrieved in step 1.
```

> 5) Map - Loop Functions State : In this state, GetFunction API is called for each Lambda ARN taken as input from previous step & status of each function is checked and appended to the output.

Reviewer: ```suggestion
5) Map - loop functions state: In this state, the GetFunction API is called for each Lambda ARN taken as input from the previous step, and the status of each function is checked and appended to the output.
```

> 6) Filter Inactive Functions State : This filters out the Active functions from the Json array and passes only the Inactive functions list to the next step.

Reviewer: ```suggestion
6) Filter inactive functions state: This filters out the active functions from the JSON array and passes only the list of inactive functions to the next step.
```

> 7) SNS Publish : List of Inactive functions is published to the SNS Topic.

Reviewer: ```suggestion
7) SNS publish : List of inactive functions is published to the SNS Topic.
```

## Cleanup
 
1. Delete the stack
    ```bash
    sam delete
    ```

2. Also, you have to manually delete any schedules you created if required.

----

> Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.

Reviewer: ```suggestion
Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
